### PR TITLE
[WIP] Commands related to script

### PR DESCRIPTION
**Pull request by kobkaz**

### CODE REVIEW

File: src/command.rs [hunk 1]
```diff
@@ -573,3 +577,169 @@ impl Command for Echo {
...
+    fn from_response(res: RespValue) -> Result<Self::Output, RespError> {
+        Ok(res)
+    }
+
+    fn key_slot(&self) -> Result<Option<u16>, Vec<u16>> {
+        hash_slot_many(&self.keys)
+    }
+}
+
+pub struct EvalSHA {
```
Comment (pandaman64) [inline src/command.rs]: Could you rename it to `EvalSha`?
Rust API Guidelines suggests not capitalizing letters in an acronym: https://rust-lang-nursery.github.io/api-guidelines/naming.html
Other crates follow this convention like [this](https://docs.rs/sha2/0.8.0/sha2/).

File: src/command.rs [hunk 2]
```diff
@@ -573,3 +577,169 @@ impl Command for Echo {
...
+pub struct ScriptFlush {
+    pub slot: u16,
+}
+
+impl Message for ScriptFlush {
+    type Result = Result<String, Error>;
+}
+
+impl Command for ScriptFlush {
+    type Output = String;
```
Comment (pandaman64) [inline src/command.rs:665]: Maybe insert a line break here for consistency?

**Comment by pandaman64** [comment]
Thank you for your contribution!
I added some comments. If you are done with the PR, please ping me.

**Review by pandaman64 — CHANGES_REQUESTED** [review]
Some reviews regarding the style.

**Comment by kobkaz** [comment]
I wrote tests for commands I am introducing in this PR. 
However, ```test_cluster_load_script``` and ```test_cluster_eval_sha``` are failing.
These failures are because the first ```SCRIPT LOAD``` request of each test is sent to a randomly selected node, as I described in #2, and the subsequent request cannot find the script which should exist in appropriate node.

File: tests/test_cluster.rs [hunk 3]
```diff
@@ -42,3 +42,117 @@ fn test_cluster_discovery() {
 
     sys.run();
 }
+
+#[test]
+fn test_cluster_eval() {
+    env_logger::try_init();
```
Comment (pandaman64) [inline tests/test_cluster.rs]: Can you move script-related tests into another file?

**Comment by pandaman64** [comment]
Thank you for adding tests!
Unfortunately, I couldn't reproduce the test failure after I replaced `env_logger::init()` in `test_cluster_discovery` with `env_logger::try_init()`.
Can you paste a backtrace with `RUST_LOG=actix_redis=debug RUST_BACKTRACE=1 cargo test`?

**Comment by kobkaz** [comment]
The test succeds with a probability of 1/3, and once it succeeds, it continues to succeed unless you reset the Redis server or explicitly issue ```SCRIPT FLUSH```.

And here's the log for a failed test.
```
$ RUST_LOG=actix_redis=debug RUS_BACKTRACE=1 cargo test test_cluster_lua_eval_sha 2>&1
   Compiling actix-redis v0.5.1 (/home/kaz/src/actix-redis-cluster)
    Finished dev [unoptimized + debuginfo] target(s) in 4.36s
     Running /home/kaz/src/actix-redis-cluster/target/debug/deps/actix_redis-6a901ba98bafe7d1

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running /home/kaz/src/actix-redis-cluster/target/debug/deps/test_cluster-53e781d7a16aa85a

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running /home/kaz/src/actix-redis-cluster/target/debug/deps/test_cluster_discovery-037fd342a70623da

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running /home/kaz/src/actix-redis-cluster/target/debug/deps/test_cluster_lua-d5fed9f1f75ef811

running 1 test
 INFO 2019-02-27T11:36:23Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-02-27T11:36:23Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-02-27T11:36:23Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: random node: 127.0.0.1:7001
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: handling [[0 5460 ["172.17.0.2" 7000 "e0b84dc619febb96b359d9c141a14d563aa3963a"] ["172.17.0.2" 7005 "49b59577c4b40c2669f34f7983a6f78de0a416d8"]] [5461 10922 ["172.17.0.2" 7001 "ee4e53cdcd8510cd5d5008b7dfb6fee2ffed9fed"] ["172.17.0.2" 7003 "35efa41c88d0a02c0b2bcf8da0534289993f14ca"]] [10923 16383 ["172.17.0.2" 7002 "31848eb78f32fd123b82ee6be313100515c9d76f"] ["172.17.0.2" 7004 "19695145cbf5a6d8af2efbb8ba19a4411f4f50e4"]]] from 127.0.0.1:7002
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: handling [[0 5460 ["172.17.0.2" 7000 "e0b84dc619febb96b359d9c141a14d563aa3963a"] ["172.17.0.2" 7005 "49b59577c4b40c2669f34f7983a6f78de0a416d8"]] [5461 10922 ["172.17.0.2" 7001 "ee4e53cdcd8510cd5d5008b7dfb6fee2ffed9fed"] ["172.17.0.2" 7003 "35efa41c88d0a02c0b2bcf8da0534289993f14ca"]] [10923 16383 ["172.17.0.2" 7002 "31848eb78f32fd123b82ee6be313100515c9d76f"] ["172.17.0.2" 7004 "19695145cbf5a6d8af2efbb8ba19a4411f4f50e4"]]] from 127.0.0.1:7001
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: handling "f42506a3b3c86c32e5eb0402829fe81220de93b2" from 127.0.0.1:7001
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: handling [[10923 16383 ["172.17.0.2" 7002 "31848eb78f32fd123b82ee6be313100515c9d76f"] ["172.17.0.2" 7004 "19695145cbf5a6d8af2efbb8ba19a4411f4f50e4"]] [5461 10922 ["172.17.0.2" 7001 "ee4e53cdcd8510cd5d5008b7dfb6fee2ffed9fed"] ["172.17.0.2" 7003 "35efa41c88d0a02c0b2bcf8da0534289993f14ca"]] [0 5460 ["172.17.0.2" 7000 "e0b84dc619febb96b359d9c141a14d563aa3963a"] ["172.17.0.2" 7005 "49b59577c4b40c2669f34f7983a6f78de0a416d8"]]] from 127.0.0.1:7000
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }]
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: [Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }]
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }]
 INFO 2019-02-27T11:36:23Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-02-27T11:36:23Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-02-27T11:36:23Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-02-27T11:36:23Z: actix_redis::cluster: handling NOSCRIPT No matching script. Please use EVAL. from 172.17.0.2:7002
test test_cluster_lua_eval_sha ... FAILED

failures:

---- test_cluster_lua_eval_sha stdout ----
thread 'test_cluster_lua_eval_sha' panicked at 'Should not happen Ok(Ok(Error("NOSCRIPT No matching script. Please use EVAL.")))', tests/test_cluster_lua.rs:115:26
note: Run with `RUST_BACKTRACE=1` for a backtrace.
Panic in Arbiter thread, shutting down system.


failures:
    test_cluster_lua_eval_sha

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 2 filtered out

error: test failed, to rerun pass '--test test_cluster_lua'
```

**Comment by pandaman64** [comment]
OK. Now I see what's wrong. Let me stop issuing commands to Redis servers during the initialization of slot info and whether this fix the problem.

**Comment by pandaman64** [comment]
I'm sorry to be that late. I ended up re-implementing `RedisClusterActor` as a synchronous actor.
This introduced a breaking change in `RedisClusterActor::start`, so please update your invocation as my review comment and test it in your local environment.

File: tests/test_cluster_lua.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+use actix::prelude::*;
+use actix_redis::{command::*, RedisClusterActor, RespValue};
+use futures::Future;
+
+#[test]
+fn test_cluster_lua_eval() {
+    let _ = env_logger::try_init();
+    let sys = System::new("test");
+
+    let addr = RedisClusterActor::start(vec![
```
Comment (pandaman64) [inline tests/test_cluster_lua.rs]: `RedisClusterActor::start` started to take the number of workers (clients), so please update this constructor as follows (I'd like to test concurrent behaviors, so please set the number of workers greater than 1):
```
RedisClusterActor::start(
    3,
    vec![...],
);
```

File: tests/test_cluster.rs [hunk 5]
```diff
@@ -45,3 +45,4 @@ fn test_cluster_discovery() {
 
     sys.run();
 }
+
```
Comment (pandaman64) [inline tests/test_cluster.rs]: Can you remove this empty line so that the formatter satisfies?

**Comment by kobkaz** [comment]
It's working fine, but when I run tests at once, I see some errors. Though tests themselves succeed, it reports panics of some threads.
Using ```cargo test --test-threads 1```  prevents this error, so it seems parallel execution of tests is the cause.
I'm not surprised if it is a fundamental limitation of using ```SyncArbiter```, but not sure if it actually is.
What do you think about this?
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running target/debug/deps/actix_redis-6a901ba98bafe7d1

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running target/debug/deps/test_cluster-53e781d7a16aa85a

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running target/debug/deps/test_cluster_discovery-037fd342a70623da

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running target/debug/deps/test_cluster_lua-d5fed9f1f75ef811

running 2 tests
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7000
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7001
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7001
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: random node: 127.0.0.1:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: [Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }]
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: [Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }]
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: [Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }]
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: [Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }]
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: [Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }]
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-03-29T05:07:06Z: actix_redis::cluster: [Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }, Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }]
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 WARN 2019-03-29T05:07:06Z: actix_redis::cluster: Failed to send a message to the resolver: MailboxError(Mailbox has closed)
test test_cluster_lua_eval_sha ... ok
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ()', src/libcore/result.rs:1009:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
 INFO 2019-03-29T05:07:06Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 WARN 2019-03-29T05:07:06Z: actix_redis::cluster: Failed to send a message to the resolver: MailboxError(Mailbox has closed)
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ()', src/libcore/result.rs:1009:5
test test_cluster_lua_eval ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running target/debug/deps/test_redis-23d32f33e9658c7d

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out

   Doc-tests actix_redis

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

```

**Comment by pandaman64** [comment]
`System::current().stop()` in a test function might terminate the actor system before the other tests complete.
One possible solution for this is just splitting each test function into separate files. If you encounter the error again, please provide the backtrace.

**Comment by kobkaz** [comment]
The error reproduces constantly, but it's backtrace seems not very informative.

```
$ RUST_BACKTRACE=full RUST_LOG=actix_redis=debug cargo test test_cluster_lua_eval -- --nocapture

Finished dev [unoptimized + debuginfo] target(s) in 0.13s
     Running target/debug/deps/actix_redis-6a901ba98bafe7d1

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running target/debug/deps/test_cluster-53e781d7a16aa85a

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running target/debug/deps/test_cluster_discovery-037fd342a70623da

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

     Running target/debug/deps/test_cluster_lua-d5fed9f1f75ef811

running 2 tests
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7001
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7002
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7000
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 127.0.0.1:7002
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7000
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7001
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: random node: 127.0.0.1:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }]
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7000
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }]
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }]
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
test test_cluster_lua_eval ... ok
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }]
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 WARN 2019-04-01T05:31:18Z: actix_redis::cluster: Failed to send a message to the resolver: MailboxError(Mailbox has closed)
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ()', src/libcore/result.rs:1009:5
stack backtrace:
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7002
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }]
 INFO 2019-04-01T05:31:18Z: actix_redis::cluster: Connected to redis server: 172.17.0.2:7001
DEBUG 2019-04-01T05:31:18Z: actix_redis::cluster: [Slots { start: 0, end: 5460, master_addr: "172.17.0.2:7000" }, Slots { start: 10923, end: 16383, master_addr: "172.17.0.2:7002" }, Slots { start: 5461, end: 10922, master_addr: "172.17.0.2:7001" }]
test test_cluster_lua_eval_sha ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out

 WARN 2019-04-01T05:31:18Z: actix_redis::cluster: Failed to send a message to the resolver: MailboxError(Mailbox has closed)
     Running target/debug/deps/test_redis-23d32f33e9658c7d

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out

   Doc-tests actix_redis

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by pandaman64** [comment]
Doesn't your `test_cluster_lua.rs` contain two test functions? The root problem, I think, is that simultaneous execution of the tests overrides the current `System`, which lead to a kind of race condition.
Both of having one test function per one file and setting `--test-threads` to one can fix this issue, though I prefer the former since the invocation of the tests can be simpler.

**Comment by kobkaz** [comment]
Sorry, forgot to push this for a week.
Please merge this PR if it's ok.

**Comment by pandaman64** [comment]
Good Job :+1:

**Pull request merged**